doc: add tidb rollup syntax (#15053)
pingcap/docs-cn: functions-and-operators/group-by-modifier.md

示例参考计划: [explain](/explain-aggregation.md#多维度数据聚合 ROLLUP)

`Expand_20` 算子信息展示了所谓生成的层级表达式：`level-projection:[test.bank.profit, <nil>->Column#6, <nil>->Column#7, 0->gid],[test.
bank.profit, Column#6, <nil>->Column#7, 1->gid],[test.bank.profit, Column#6, Column#7, 3->gid]`。其由 2 维表达式组成，尾部后缀有 
`Expand` 算子的 Schema 信息：`schema: [test.bank.profit,Column#6,Column#7,gid]`。

正上所示，在 `Expand` 算子的 Schema 信息中，`GID` 会作为额外的生成列来输出，其值也是由 `Expand` 算子根据不同的维度分组逻辑计算得来的，反应了当前数据副本和维度分组的关系。其中最常见的属是位掩码运算, 其可以容纳 63 种 GROUP BY ITEMS 的 ROLLUP 组合，对应维度分组的数量为 64 种。这种模式下 `GID` 值的生成根据当前数据副本复制时所需维度分组中 GROUPING EXPRESSION 的有无，按照 GROUP BY ITEMS 的序列，顺序填充一个 64 位的 UINT64 的值。

例如这里 GROUP BY ITEMS 序列顺序是 [year, month], 而 ROLLUP 的语法生成的维度分组集合为：{year, month}, {year}, {}。对于维度分组 {year, month} 来说，其 `year` 和 `month` 两个 GROUPING ITEM 都是当前维度分组所需的列，对应填充对应比特位为 1 和 1，组成 UINT64 为 11...0 即 3。相应的 `year` 和 `month` 两个 GROUPING ITEM 都是当前维度分组所需的列，所在当前投影表达式构建时要保留这两个列的完整输出，所以投影表达式为 `[test.bank.profit, Column#6, Column#7, 3->gid]`。（column#6 = year, column#7 = month）

> 

Reviewer: 我觉得在这里之后举一个具体的例子，比如某一行数据
```
10.3  2000   'Jan'
```
经过 Expand 之后会生成以下 3 行数据
```
10.3  NULL  NULL 0
10.3  2000  NULL 1
10.3  2000  'Jan' 3
```

---

> title: GROUP BY 修饰符

Reviewer 2: 对于新增文档，需要加到文档目录里，请修改下 TOC.md 以及 TOC-cloud.md

> summary: 关于 TiDB GROUP BY 修饰符的使用

Reviewer 2: ```suggestion
summary: 了解如何使用 TiDB GROUP BY 修饰符。
```

> aliases: ['/docs/dev/functions-and-operators/group-by-modifiers/','/docs/dev/reference/sql/functions-and-operators/group-by-modifiers/']

Reviewer 2: ```suggestion
```

---

> # GROUP BY Modifiers

Reviewer 2: ```suggestion
# GROUP BY 修饰符
```

> ## 普通使用指南

Reviewer 3: 前面或者最后空个位置说下不支持 CUBE 之类的限制？

> 从 v7.4 开始，TiDB GROUP-BY 条项允许后接 WITH ROLLUP 修饰符来对 GROUP BY ITEMS 进行多个分组维度的描述，从而使得聚合输出是能够包含在各个分组维度上聚合结果的并集。ROLLUP 的语法规定，对前缀 GROUP-BY 所包含的分组表达式列表，依次递减列表尾端的分组表达式元素，每次递减一项，形成新的分组高维度序列，该语句所有上层执行流需要在新的分组维度上也进行依次计算，并将输出结果和前序其他聚合结果进行并集累加；分组表达式列表最终可以递减到空集序列，语义上意味着聚合逻辑需要在全域数据范围进行计算，最终的输出结果依旧是和前序其他聚合结果进行并集累加，最终统一输出到客户端。

Reviewer 2: ```suggestion
自 v7.4.0 起，TiDB  的 `GROUP BY` 子句支持 `WITH ROLLUP` 修饰符。当你在 `GROUP BY` 子句中添加 `WITH ROLLUP` 修饰符时，TiDB 可以对 `GROUP BY` 的列进行多层次的分组，并在输出中为你提供各个分组的汇总结果。

`WITH ROLLUP` 的语法要求按照 `GROUP BY` 子句中列出的分组表达式递减地生成新的分组维度序列。查询需要在新的分组维度上进行计算，并将结果与之前的聚合结果进行并集累加。分组表达式列表最终可以递减为空集序列，这意味着聚合逻辑需要在全域数据范围内进行计算，最终的输出结果依旧是和前序其他聚合结果进行并集累加，最终统一输出到客户端。
```

```sql
select count(1) from t GROUP BY a,b,c WITH ROLLUP;

> -- 其意味着，count(1) 的结果需要数据分别在，{a,b,c}, {a,b}, {a},{} 一共 N+1 个分组上进行聚合，然后联合输出（考虑 GROUP BY ITEMS 的长度为 N）。

Reviewer 2: ```suggestion
-- 在此示例中，count(1) 的结果需要数据分别在，{a,b,c}, {a,b}, {a},{} 一共 N+1 个分组上进行聚合，然后联合输出（考虑 GROUP BY ITEMS 的长度为 N）。
```

```

> 高维度聚合联合输出一般是用在 OLAP（Online Analytical Processing）场景下居多，因此目前，TiDB 仅在 MPP 模式下才支持 对ROLLUP 语法产生有效的执行计划，这也意味客户集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。目前 ROLLUP 的实现引入了一个新的算子 Expand，该算子能够根据不同的分组规则，进行底层数据的特殊复制，不同复制的数据份对应于一个特定的分组规则；利用 TiDB MPP 模式下对 Expand 之后的大批量数据的灵活 shuffle 来进行实现高效分组和聚合计算，均摊多节点算力。

Reviewer 2: ```suggestion
## 使用场景

高维度的聚合联合输出一般常用于 OLAP（Online Analytical Processing）场景。目前 ROLLUP 的实现引入了一个新的算子 Expand，该算子能够根据不同的分组规则，进行底层数据的特殊复制，不同复制的数据份对应于一个特定的分组规则；利用 TiDB MPP 模式下对 Expand 之后的大批量数据的灵活 shuffle 来进行实现高效分组和聚合计算，均摊多节点算力。

## 准备条件

目前，TiDB 仅在 MPP 模式下才支持 对 ROLLUP 语法产生有效的执行计划，这也意味你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。

更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。

## 使用示例

```

|               └─Projection_16          | 3.00    | mpp[tiflash] |               | test.bank.profit, test.bank.year->Column#6, test.bank.month->Column#7                                                                                                                                                                |
|                 └─TableFullScan_17     | 3.00    | mpp[tiflash] | table:bank    | keep order:false, stats:pseudo                                                                                                                                                                                                       |
+----------------------------------------+---------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
10 rows in set (0.05 sec)

> ```

Author: 好的

2 rows in set (0.15 sec)
```
对于银行报表来说，有时候出来分析计算每年的利润之后，通常我们还会算下所有年份的利润综合，甚至每个月的利润总成来达到高维度或者是更细粒度的利润分析；然后这并不是一个简单的 GROUP-BY 语句能做的，通常是要多个 GROUP-BY 语句跟上不同的聚合粒度，然后将所有的所有用 UNION 链接起来呈现。在有了 ROLLUP 语法之后，其实可以这样做：
```sql
tidb> SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP;
+------+-------+--------------------+
| year | month | profit             |
+------+-------+--------------------+
| 2000 | Jan   | 10.300000190734863 |
| 2001 | NULL  | 22.399999618530273 |
| 2000 | NULL  |  41.90000057220459 |
| 2001 | Feb   | 22.399999618530273 |
| 2000 | Mar   | 31.600000381469727 |
| NULL | NULL  |  64.30000019073486 |
+------+-------+--------------------+
6 rows in set (0.05 sec)
```
在这份结果中，联合了在不同的月份，年份，以及全域上的所有维度的聚合输出；其中 month 和 year 部分出现的 `NULL` 值表示这个 month 或 year 是被高维度聚合分组所 GROUP 掉的列，该行 profit 所呈现的聚合结果 profit，是仅在没有被 `NULL` 化的 month 和 year 所组成的名文所对应的分组上所聚合出来的。

简单就示例来说：
* 第一行的 profit 结果是来源于在 year 和 month 都没有被 GROUP 掉的情况下，并来自 2 维分组 {year, month} 且其具体值为 {2000, “Jan”} 的细粒度小分组的聚合结果。

* 第二行的 profit 结果是来源于仅在 month 被 GROUP 掉的情况下，来自 1 维分组 {year} 且其具体值为 {2001} 的中层粒度分组下的聚合结果。

* 最后一行的 profit 结果是来源于在 year 和 month 都被 GROUP 掉的情况下，来自 0 维分组 {} 即在全域范围内展开的结果聚合。

> 考虑到这种 `NULL` 有特殊的含义 --- 分组表达式所呈现出来的 `NULL` 值可以被认为是在高维度聚合中被 GROUP 掉的那一列，因此使用整个分组转向于更高维度的聚合粒度。这种 `NULL` 值的赋予动作是 Expand 算子在复制逻辑中添加的，而 Expand 算子的构建就在 Aggregate 算子的前序动作当中，因此该分组表达式所呈现出来的 `NULL` 可以被逻辑 Aggregate 算子之后的所有其他子句所利用，主要表现为 select-list / having / order-by 当中表达式的应用。 比如：我们可以利用 having 子句过滤并只看 2 维度分组下的聚合结果输出。

Reviewer 2: ```suggestion
`WITH ROLLUP` 结果中的  `NULL` 值是在应用 Aggregate 算子之前生成的，因此你可以将 `NULL` 值应用于 `SELECT`、`HAVING`、`ORDER BY` 子句中，进一步过滤聚合结果。

例如，你可以在 `HAVING` 子句中通过 `NULL` 过滤并只看 2 维度分组下的聚合结果输出。
```

```sql

> tidb> SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP having year is not null and month is not null;

Reviewer 2: ```suggestion
SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP having year is not null and month is not null;
```

```

也是考虑到这种分组表达式所呈现出来的 `NULL` 的特殊含义，如果分组表达式原有数据本身就包含有原生 `NULL` 值，那么可能会影响我们对聚合结果所在的分组粒度上有所误判。因为为了区分更好的区别这两种 `NULL` 值的来源，我们配套引入了 `GROUPING()` 函数来接受分组表达式作为参数，并输出 `0` 或者 `1` 表示该分组表达式在当前聚合结果输出行中是否被 GROUP 掉了。`1` 表示是，意味着该分组聚合结果转向了更高维的聚合，而 `0` 则反之。

```sql
tidb> SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(month) as grp_month from bank GROUP BY year, month WITH ROLLUP;
+------+-------+--------------------+----------+-----------+
| year | month | profit             | grp_year | grp_month |
+------+-------+--------------------+----------+-----------+
| 2000 | Jan   | 10.300000190734863 |        0 |         0 |
| 2000 | Mar   | 31.600000381469727 |        0 |         0 |
| NULL | NULL  |  64.30000019073486 |        1 |         1 |
| 2000 | NULL  |  41.90000057220459 |        0 |         1 |
| 2001 | Feb   | 22.399999618530273 |        0 |         0 |
| 2001 | NULL  | 22.399999618530273 |        0 |         1 |
+------+-------+--------------------+----------+-----------+
6 rows in set (0.07 sec)
```

可以看到我们不再依赖分组表达式 year 和 month 的输出去判断该聚合结果行所在的聚合维度，而是依赖于 grp_year 和 grp_month 的 GROUPING 函数结果来判断，以防止分组表达式 year 和 month 自有原生 `NULL` 值的干扰。

> `GROUPING()` 函数最大可以接受 64 个分组表达式作为参数，而多参数的他们输出不再是简单的 `0` 和 `1`, 而是每个参数都可以生成一个 `0` 或 `1` 的结果，综合组成一个比特位的 `0` 或 `1` 总体是 64 位的 `UNSIGNED LONGLONG`，而其所在该比特数位中的位置可以通过以下公式计算：

Reviewer 2: ```suggestion
`GROUPING()` 函数最多可以接受 64 个分组表达式作为参数。在多参数的输出中，每个参数都可以生成一个 `0` 或 `1` 的结果，综合组成一个比特位的 `0` 或 `1` 总体是 64 位的 `UNSIGNED LONGLONG`。而其所在该比特数位中的位置可以通过以下公式计算：
```

```go
GROUPING(day, month, year):
  result for GROUPING(year)
+ result for GROUPING(month) << 1
+ result for GROUPING(day) << 2
```

使用组合参数的 `GROUPING()` 的函数可以快速过滤掉出任何高维度的聚合结果，就像这样只查看高维度聚合的结果：
```sql

> tidb> SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(month) as grp_month from bank GROUP BY year, month WITH ROLLUP having GROUPING(year, month) <> 0;

Reviewer 2: ```suggestion
SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(month) as grp_month from bank GROUP BY year, month WITH ROLLUP having GROUPING(year, month) <> 0;
```

+------+-------+--------------------+----------+-----------+
| year | month | profit             | grp_year | grp_month |
+------+-------+--------------------+----------+-----------+
| 2000 | NULL  |  41.90000057220459 |        0 |         1 |
| NULL | NULL  |  64.30000019073486 |        1 |         1 |
| 2001 | NULL  | 22.399999618530273 |        0 |         1 |
+------+-------+--------------------+----------+-----------+
3 rows in set (0.11 sec)
```

> ## 如何阅读 ROLLUP 的执行计划

Reviewer: 这种实现细节需要写在用户文档里面吗？

Author: @winoros yiding 觉得呢，要加计划阅读解释吗？

Reviewer 3: 可以把代码的部分删了吧 简单说一下 Expand 的 expression 是干啥用的（

Reviewer 4: 我觉得有点太细节了。不用介绍实现细节，只要能够让用户理解计划中对应的含义就够了。

Author: 好的

| 2001 | Feb   | 22.399999618530273 |
+------+-------+--------------------+
3 rows in set (0.02 sec)
```

需要注意的是，如果 `GROUP BY` 分组列表中的某列包含原生的 `NULL` 值，`WITH ROLLUP` 的分组聚合可能会对查询结果产生误导。为了解决这个问题，你可以使用 `GROUPING ()` 函数区分原生的 `NULL` 值和 `WITH ROLLUP` 生成的 `NULL` 值。该函数接受分组表达式作为参数，并输出 `0` 或 `1`，表示该分组表达式是否在当前结果中被聚合。`1` 表示被聚合，`0` 表示没有。

以下是如何使用 `GROUPING ()` 函数的示例：

```sql
SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(month) as grp_month FROM bank GROUP BY year, month WITH ROLLUP ORDER BY year DESC, month DESC;
+------+-------+--------------------+----------+-----------+
| year | month | profit             | grp_year | grp_month |
+------+-------+--------------------+----------+-----------+
| 2001 | Feb   | 22.399999618530273 |        0 |         0 |
| 2001 | NULL  | 22.399999618530273 |        0 |         1 |
| 2000 | Mar   | 31.600000381469727 |        0 |         0 |
| 2000 | Jan   | 10.300000190734863 |        0 |         0 |
| 2000 | NULL  |  41.90000057220459 |        0 |         1 |
| NULL | NULL  |  64.30000019073486 |        1 |         1 |
+------+-------+--------------------+----------+-----------+
6 rows in set (0.028 sec)

```

> 可以看到我们不再依赖分组表达式 year 和 month 的输出去判断该聚合结果行所在的聚合维度，而是依赖于 grp_year 和 grp_month 的 GROUPING 函数结果来判断，以防止分组表达式 year 和 month 自有原生 `NULL` 值的干扰。

Reviewer 2: ```suggestion
在此输出中，你可以直接通过 `grp_year` 和 `grp_month` 的结果来判断该聚合结果行所在的聚合维度，以防止分组表达式 `year` 和 `month` 原生的 `NULL` 值的干扰。
```

> `GROUPING()` 函数最多可以接受 64 个分组表达式作为参数。在多参数的输出中，每个参数都可以生成一个 `0` 或 `1` 的结果，综合组成一个比特位的 `0` 或 `1` 总体是 64 位的 `UNSIGNED LONGLONG`。而其所在该比特数位中的位置可以通过以下公式计算：

Reviewer 2: ```suggestion
`GROUPING()` 函数最多可以接受 64 个分组表达式作为参数。在多参数的输出中，每个参数都可以生成一个 `0` 或 `1` 的结果。综合组成每一个比特位是 `0` 或 `1` 最多 64 位的 `UNSIGNED LONGLONG`。各个参数在比特数位中的位置可以通过以下公式计算：
```

Reviewer 2: 这里不太确定修改的是否准确

```go
GROUPING(day, month, year):
  result for GROUPING(year)
+ result for GROUPING(month) << 1
+ result for GROUPING(day) << 2
```

> 使用组合参数的 `GROUPING()` 的函数可以快速过滤掉出任何高维度的聚合结果，就像这样只查看高维度聚合的结果：

Reviewer 2: ```suggestion
在 `GROUPING()` 的函数中使用组合参数可以快速过滤出任何高维度的聚合结果。例如，你可以通过 `GROUPING(year, month)` 快速过滤出每年以及所有年份的聚合结果：

```sql
SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(month) as grp_month FROM bank GROUP BY year, month WITH ROLLUP HAVING GROUPING(year, month) <> 0 ORDER BY year DESC, month DESC;
+------+-------+--------------------+----------+-----------+
| year | month | profit             | grp_year | grp_month |
+------+-------+--------------------+----------+-----------+
| 2001 | NULL  | 22.399999618530273 |        0 |         1 |
| 2000 | NULL  |  41.90000057220459 |        0 |         1 |
| NULL | NULL  |  64.30000019073486 |        1 |         1 |
+------+-------+--------------------+----------+-----------+
3 rows in set (0.023 sec)

> 

Reviewer 2: ```suggestion
```

`Expand_20` 算子信息展示了所谓生成的层级表达式：`level-projection:[test.bank.profit, <nil>->Column#6, <nil>->Column#7, 0->gid],[test.bank.profit, Column#6, <nil>->Column#7, 1->gid],[test.bank.profit, Column#6, Column#7, 3->gid]`。其由 2 维表达式组成，尾部后缀有 `Expand` 算子的 Schema 信息：`schema: [test.bank.profit,Column#6,Column#7,gid]`。

正上所示，在 `Expand` 算子的 Schema 信息中，`GID` 会作为额外的生成列来输出，其值也是由 `Expand` 算子根据不同的维度分组逻辑计算得来的，反应了当前数据副本和维度分组的关系。其中最常见的属是位掩码运算, 其可以容纳 63 种 GROUP BY ITEMS 的 ROLLUP 组合，对应维度分组的数量为 64 种。这种模式下 `GID` 值的生成根据当前数据副本复制时所需维度分组中 GROUPING EXPRESSION 的有无，按照 GROUP BY ITEMS 的序列，顺序填充一个 64 位的 UINT64 的值。

例如这里 GROUP BY ITEMS 序列顺序是 [year, month], 而 ROLLUP 的语法生成的维度分组集合为：{year, month}, {year}, {}。对于维度分组 {year, month} 来说，其 `year` 和 `month` 两个 GROUPING ITEM 都是当前维度分组所需的列，对应填充对应比特位为 1 和 1，组成 UINT64 为 11...0 即 3。相应的 `year` 和 `month` 两个 GROUPING ITEM 都是当前维度分组所需的列，所在当前投影表达式构建时要保留这两个列的完整输出，所以投影表达式为 `[test.bank.profit, Column#6, Column#7, 3->gid]`。（column#6 = year, column#7 = month）

```
以原始数据中的一行为例：
+------+-------+------+------------+
| year | month | day  | profit     |
+------+-------+------+------------+
| 2000 | Jan   |    1 | 10.3000000 |
+------+-------+------+------------+

> 经过 Expland 算子之后，可以得到以下三行结果：

Author: ```suggestion
经过 Expand 算子之后，可以得到以下三行结果：
```

> 根据 Projection 表达式投影之后，可以得到以下三行结果：

Reviewer: ```suggestion
经过 Expland 算子之后，可以得到以下三行结果：
```

---
title: GROUP BY 修饰符
summary: 了解如何使用 TiDB GROUP BY 修饰符。
---
# GROUP BY 修饰符

自 v7.4.0 起，TiDB 的 `GROUP BY` 子句支持 `WITH ROLLUP` 修饰符。

> 你可以在 `GROUP BY` 子句中指定一个或多个列，形成一个分组列表，然后添加 `WITH ROLLUP` 修饰符。TiDB 将会按照分组列表中的列进行多层次的递减分组，并在输出中为你提供各个分组数据的汇总结果。
> 
> - 分组方式：
>     - 第一个分组层次为分组表列表中的所有列。
>     - 后面的层次将从分组列表的最右侧（尾端）开始，每次递减一个元素，形成新的分组。
> - 聚合汇总：在每个层次上，查询都会执行聚合操作，然后将该层次的计算结果与前面所有层次的结果进行汇总。这意味着你可以看到不同级别的聚合数据，从详细到总体。

Reviewer 2: ```suggestion
你可以在 `GROUP BY` 子句中指定一个或多个列，形成一个分组列表，然后添加 `WITH ROLLUP` 修饰符。TiDB 将会按照分组列表中的列进行多维度的递减分组，并在输出中为你提供各个分组数据的汇总结果。

- 分组方式：
    - 第一个分组维度为分组表列表中的所有列。
    - 后面的维度将从分组列表的最右侧（尾端）开始，每次递减一个元素，形成新的分组。
- 聚合汇总：在每个维度上，查询都会执行聚合操作，然后将该维度的计算结果与前面所有维度的结果进行汇总。这意味着你可以看到不同维度的聚合数据，从详细到总体。
```

Reviewer 2: 不确定，“层次”改成“维度”是否更准确

-- 在此示例中，count(1) 的结果需要数据分别在，{a,b,c}, {a,b}, {a},{} 一共 N+1 个分组上进行聚合，然后联合输出分组的汇总数据（考虑 GROUP BY ITEMS 的长度为 N）。
```
## 使用场景

多列数据的聚合汇总输出一般常用于 OLAP（Online Analytical Processing）场景。目前 ROLLUP 的实现引入了一个新的算子 Expand，该算子能够根据不同的分组规则，进行底层数据的特殊复制，不同复制的数据份对应于一个特定的分组规则；利用 TiDB MPP 模式下对 Expand 之后的大批量数据的灵活 shuffle 来进行实现高效分组和聚合计算，均摊多节点算力。

## 准备条件

目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。

更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。

## 使用示例

假如有一张名为 `bank` 的银行利润表，包含年（`year`）、月（`month`）、日（`day`）和利润（`profit`）列。

```sql
CREATE TABLE bank
(
    year    INT,
    month   VARCHAR(32),
    day     INT,
    profit  FLOAT
);

ALTER TABLE bank set SET TIFLASH REPLICA 1; -- 为该表添加一个 TiFlash 副本

INSERT INTO bank VALUES(2000, "Jan", 1, 10.3),(2001, "Feb", 2, 22.4),(2000,"Mar", 3, 31.6)
```
如需查看银行每年的利润，可以用一个简单的 `GROUP BY` 的子句来实现：

```sql
SELECT year, SUM(profit) AS profit FROM bank GROUP BY year;
+------+--------------------+
| year | profit             |
+------+--------------------+
| 2001 | 22.399999618530273 |
| 2000 |  41.90000057220459 |
+------+--------------------+
2 rows in set (0.15 sec)
```

> 对于银行报表来说，除了每年的利润之外，通常还需要计算所有年份的总利润，甚至每个月的总利润，以进行更高层次或更详细的利润分析。在 v7.4.0 之前的版本中，你需要多次使用不同的 `GROUP BY` 子句，并将结果使用 UNION 连接，才能得到聚合汇总的结果。从 v7.4.0 起，你可以直接在 `GROUP BY` 子句中添加 `WITH ROLLUP` 修饰符，即可得到所需的结果：

Reviewer 2: ```suggestion
对于银行报表来说，除了每年的利润之外，通常还需要计算所有年份的总利润和每个月的总利润，以进行更高层次或更详细的利润分析。在 v7.4.0 之前的版本中，你需要多次使用不同的 `GROUP BY` 子句，并将结果使用 UNION 连接，才能得到聚合汇总的结果。从 v7.4.0 起，你可以直接在 `GROUP BY` 子句中添加 `WITH ROLLUP` 修饰符，即可得到所需的结果：
```

```sql

> TiDB [test]> SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP order by year desc, month desc;

Reviewer 2: ```suggestion
SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP order by year desc, month desc;
```

10 rows in set (0.05 sec)

在这个示例执行计划中，你可以在 `Expand_20` 这行查看`Expand` 算子的层级表达式，其由 2 维表达式组成，行末有 `Expand` 算子的 Schema 信息 `schema: [test.bank.profit,Column#6,Column#7,gid]`。

在 `Expand` 算子的 schema 信息中，`GID` 会作为额外的生成列来输出，其值是由 `Expand` 算子根据不同维度的分组逻辑计算得出，反映了当前数据副本和维度分组的关系。最常见的情况是使用位掩码运算, 它可以表示 63 种分组项的 ROLLUP 组合，对应 64 种维度的分组。在这种模式下，`GID` 值的生成根据当前数据副本复制时所需维度分组中是否有分组表达式，按照要进行分组的列，顺序填充一个 64 位的 UINT64 的值。

例如这里 GROUP BY ITEMS 序列顺序是 [year, month], 而 ROLLUP 的语法生成的维度分组集合为：{year, month}, {year}, {}。对于维度分组 {year, month} 来说，其 `year` 和 `month` 两个 GROUPING ITEM 都是当前维度分组所需的列，对应填充对应比特位为 1 和 1，组成 UINT64 为 11...0 即 3。相应的 `year` 和 `month` 两个 GROUPING ITEM 都是当前维度分组所需的列，所在当前投影表达式构建时要保留这两个列的完整输出，所以投影表达式为 `[test.bank.profit, Column#6, Column#7, 3->gid]`。（column#6 = year, column#7 = month）

```
以原始数据中的一行为例：
+------+-------+------+------------+
| year | month | day  | profit     |
+------+-------+------+------------+
| 2000 | Jan   |    1 | 10.3000000 |
+------+-------+------+------------+

> 
> 经过 Expand 算子之后，可以得到以下三行结果：

Reviewer 2: ````suggestion
```

经过 Expand 算子之后，可以得到以下三行结果：
```sql
````

|                 └─TableFullScan_17     | 3.00    | mpp[tiflash] | table:bank    | keep order:false, stats:pseudo                                                                                                                                                                                                       |
+----------------------------------------+---------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
10 rows in set (0.05 sec)

在这个示例执行计划中，你可以在 `Expand_20` 这行查看`Expand` 算子的层级表达式，其由 2 维表达式组成，行末有 `Expand` 算子的 Schema 信息 `schema: [test.bank.profit,Column#6,Column#7,gid]`。

在 `Expand` 算子的 schema 信息中，`GID` 会作为额外的生成列来输出，其值是由 `Expand` 算子根据不同维度的分组逻辑计算得出，反映了当前数据副本和维度分组的关系。最常见的情况是使用位掩码运算, 它可以表示 63 种分组项的 ROLLUP 组合，对应 64 种维度的分组。在这种模式下，`GID` 值的生成根据当前数据副本复制时所需维度分组中是否有分组表达式，按照要进行分组的列，顺序填充一个 64 位的 UINT64 的值。

> 例如，这里分组列表中列的顺序是 [year, month]，而 ROLLUP 语法生成的维度分组集合为：{year, month}, {year}, {}。对于维度分组 {year, month} 来说，`year` 和 `month` 都是当前维度分组所需的列，对应填充比特位 1 和 1，组成 UINT64 为 11...0 即 3，因投影表达式为 `[test.bank.profit, Column#6, Column#7, 3->gid]`。（column#6 对应 year，column#7对应 month）

Reviewer 2: ```suggestion
例如，这里分组列表中列的顺序是 [year, month]，而 ROLLUP 语法生成的维度分组集合为：{year, month}, {year}, {}。对于维度分组 {year, month} 来说，`year` 和 `month` 都是当前维度分组所需的列，对应填充比特位 1 和 1，组成 UINT64 为 11...0 即 3，因此投影表达式为 `[test.bank.profit, Column#6, Column#7, 3->gid]`。（column#6 对应 year，column#7对应 month）
```

|               └─Projection_16          | 3.00    | mpp[tiflash] |               | test.bank.profit, test.bank.year->Column#6, test.bank.month->Column#7                                                                                                                                                                |
|                 └─TableFullScan_17     | 3.00    | mpp[tiflash] | table:bank    | keep order:false, stats:pseudo                                                                                                                                                                                                       |
+----------------------------------------+---------+--------------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
10 rows in set (0.05 sec)
```

> 在这个示例执行计划中，你可以在 `Expand_20` 这行查看`Expand` 算子的层级表达式，其由 2 维表达式组成，行末有 `Expand` 算子的 Schema 信息 `schema: [test.bank.profit,Column#6,Column#7,gid]`。

Reviewer 2: ```suggestion
在这个示例执行计划中，你可以在 `Expand_20` 这行的 `operator info` 列查看 `Expand` 算子的层级表达式，其由 2 维表达式组成，行末有 `Expand` 算子的 Schema 信息 `schema: [test.bank.profit,Column#6,Column#7,gid]`。
```

---
title: GROUP BY 修饰符
summary: 了解如何使用 TiDB GROUP BY 修饰符。
---

> # GROUP BY 修饰符

Reviewer 5: ```suggestion

# GROUP BY 修饰符
```

自 v7.4.0 起，TiDB 的 `GROUP BY` 子句支持 `WITH ROLLUP` 修饰符。

你可以在 `GROUP BY` 子句中指定一个或多个列，形成一个分组列表，然后添加 `WITH ROLLUP` 修饰符。TiDB 将会按照分组列表中的列进行多维度的递减分组，并在输出中为你提供各个分组数据的汇总结果。

- 分组方式：
    - 第一个分组维度为分组表列表中的所有列。
    - 后面的维度将从分组列表的最右侧（尾端）开始，每次递减一个元素，形成新的分组。
- 聚合汇总：在每个维度上，查询都会执行聚合操作，然后将该维度的计算结果与前面所有维度的结果进行汇总。这意味着你可以看到不同维度的聚合数据，从详细到总体。

> 按照这种分组方式，当分组列表中有 N 个列时，查询的计算结果将会在 N+1 个分组上进行聚合后输出。

Reviewer 5: ```suggestion
按照这种分组方式，当分组列表中有 N 个列时，查询的计算结果将会在第 N+1 个分组上进行聚合后输出。
```

多列数据的聚合汇总输出一般常用于 OLAP（Online Analytical Processing）场景。通过使用 `WITH ROLLUP` 修饰符，你可以在聚合结果中得到额外的行，以展示不同维度的汇总信息，从而实现高级的数据分析和报表生成。

## 准备条件

目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。

更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。

## 使用示例

假如有一张名为 `bank` 的银行利润表，包含年（`year`）、月（`month`）、日（`day`）和利润（`profit`）列。

```sql
CREATE TABLE bank
(
    year    INT,
    month   VARCHAR(32),
    day     INT,
    profit  DECIMAL(13, 7)
);

ALTER TABLE bank SET TIFLASH REPLICA 1; -- 为该表添加一个 TiFlash 副本

INSERT INTO bank VALUES(2000, "Jan", 1, 10.3),(2001, "Feb", 2, 22.4),(2000,"Mar", 3, 31.6)
```

如需查看银行每年的利润，可以用一个简单的 `GROUP BY` 的子句来实现：

```sql
SELECT year, SUM(profit) AS profit FROM bank GROUP BY year;
+------+--------------------+
| year | profit             |
+------+--------------------+
| 2001 | 22.399999618530273 |
| 2000 |  41.90000057220459 |
+------+--------------------+
2 rows in set (0.15 sec)
```

对于银行报表来说，除了每年的利润之外，通常还需要计算所有年份的总利润和每个月的总利润，以进行更高层次或更详细的利润分析。在 v7.4.0 之前的版本中，你需要在多个查询中使用不同的 `GROUP BY` 子句，并将结果使用 UNION 连接，才能得到聚合汇总的结果。从 v7.4.0 起，你可以直接在单个查询的 `GROUP BY` 子句中添加 `WITH ROLLUP` 修饰符，即可得到所需的结果：

```sql

> SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP order by year desc, month desc;

Reviewer 5: ```suggestion
SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP ORDER BY year desc, month desc;
```

+------+-------+--------------------+
| year | month | profit             |
+------+-------+--------------------+
| 2001 | Feb   | 22.399999618530273 |
| 2001 | NULL  | 22.399999618530273 |
| 2000 | Mar   | 31.600000381469727 |
| 2000 | Jan   | 10.300000190734863 |
| 2000 | NULL  |  41.90000057220459 |
| NULL | NULL  |  64.30000019073486 |
+------+-------+--------------------+
6 rows in set (0.025 sec)
```

以上结果包含了按照年份、月份、以及整体所有维度的聚合数据。其中，未出现 `NULL` 值的行表示该行 profit 是按照年份和月份分组计算的结果，`month` 列的 `NULL` 值表示该行 profit 是按照月份聚合计算的结果，`year` 列的 `NULL` 值表示该行 profit 是按照年份聚合计算的结果，

具体来说：

> * 第一行的 profit 值来自 2 维分组 {year, month}，为 {2000, “Jan”} 的细粒度分组的聚合结果。
> * 第二行的 profit 值来自 1 维分组 {year}，为 {2001} 的中层粒度分组下的聚合结果。
> * 最后一行的 profit 值来自 0 维分组 {}，即整体的聚合结果。

Reviewer 5: ```suggestion
* 第一行的 `profit` 值来自 2 维分组 {year, month}，为 {2000, “Jan”} 的细粒度分组的聚合结果。
* 第二行的 `profit` 值来自 1 维分组 {year}，为 {2001} 的中层粒度分组下的聚合结果。
* 最后一行的 `profit` 值来自 0 维分组 {}，即整体的聚合结果。
```

  result for GROUPING(year)
+ result for GROUPING(month) << 1
+ result for GROUPING(day) << 2
```

在 `GROUPING()` 的函数中使用组合参数可以快速过滤出任何高维度的聚合结果。例如，你可以通过 `GROUPING(year, month)` 快速过滤出每年以及所有年份的聚合结果：

```sql
SELECT year, month, SUM(profit) AS profit, grouping(year) as grp_year, grouping(month) as grp_month FROM bank GROUP BY year, month WITH ROLLUP HAVING GROUPING(year, month) <> 0 ORDER BY year DESC, month DESC;
+------+-------+--------------------+----------+-----------+
| year | month | profit             | grp_year | grp_month |
+------+-------+--------------------+----------+-----------+
| 2001 | NULL  | 22.399999618530273 |        0 |         1 |
| 2000 | NULL  |  41.90000057220459 |        0 |         1 |
| NULL | NULL  |  64.30000019073486 |        1 |         1 |
+------+-------+--------------------+----------+-----------+
3 rows in set (0.023 sec)
```

## 如何阅读 ROLLUP 的执行计划

多维度数据聚合使用了 `Expand` 算子来复制数据以满足多维度分组的需求，每个复制的数据副本都对应一个特定维度的分组。通过 MPP 的数据 shuffle 能力，`Expand` 算子能够快速地在多个 TiFlash 节点之间重新组织和计算大量的数据，充分利用每个节点的计算能力。

`Expand` 算子的实现类似 `Projection` 算子，但区别在于 `Expand` 是多层级的 `Projection`，具有多层级投影运算表达式。对于每行原始数据行，`Projection` 算子只会生成一行结果输出，而 `Expand` 算子会生成多行结果（行数等于多层级投影运算表达式的层数）。

> 以下为一个示例执行计划：

Reviewer 5: ```suggestion
以下为一个执行计划示例：
```

10 rows in set (0.05 sec)
```

在这个示例执行计划中，你可以在 `Expand_20` 这行查看`Expand` 算子的层级表达式，其由 2 维表达式组成，行末有 `Expand` 算子的 Schema 信息 `schema: [test.bank.profit,Column#6,Column#7,gid]`。

在 `Expand` 算子的 schema 信息中，`GID` 会作为额外的生成列来输出，其值是由 `Expand` 算子根据不同维度的分组逻辑计算得出，反映了当前数据副本和维度分组的关系。最常见的情况是使用位掩码运算, 它可以表示 63 种分组项的 ROLLUP 组合，对应 64 种维度的分组。在这种模式下，`GID` 值的生成根据当前数据副本复制时所需维度分组中是否有分组表达式，按照要进行分组的列，顺序填充一个 64 位的 UINT64 的值。

例如，这里分组列表中列的顺序是 [year, month]，而 ROLLUP 语法生成的维度分组集合为：{year, month}, {year}, {}。对于维度分组 {year, month} 来说，`year` 和 `month` 都是当前维度分组所需的列，对应填充比特位 1 和 1，组成 UINT64 为 11...0 即 3，因此投影表达式为 `[test.bank.profit, Column#6, Column#7, 3->gid]`。（`column#6` 对应 `year`，`column#7` 对应 `month`）

以原始数据中的下面这行为例：

```sql
+------+-------+------+------------+
| year | month | day  | profit     |
+------+-------+------+------------+
| 2000 | Jan   |    1 | 10.3000000 |
+------+-------+------+------------+
```

> 经过 Expand 算子之后，可以得到以下三行结果：

Reviewer 2: ```suggestion
经过 `Expand` 算子之后，可以得到以下三行结果：
```

```sql
+------------+------+-------+-----+ 
| profit     | year | month | gid | 
+------------+------+-------+-----+
| 10.3000000 | 2000 | Jan   |  3  | 
+------------+------+-------+-----+ 
| 10.3000000 | 2000 | NULL  |  1  | 
+------------+------+-------+-----+ 
| 10.3000000 | NULL | NULL  |  0  | 
+------------+------+-------+-----+
```

> `SELECT` 子句中使用了 `GROUPING` 函数。当在 `SELECT`、`HAVING`、`ORDER BY` 子句中使用 `GROUPING` 函数时，TiDB 会在逻辑优化阶段对其进行改写，将 `GROUPING` 函数与分组项（`GROUP BY` items）之间的关系，转化为与维度分组计算逻辑有关的 `GID`，并将此 `GID` 以 metadata 形式填充到新的 `GROUPING` 函数当中。

Reviewer 2: ```suggestion
需要注意的是，该查询的 `SELECT` 子句中使用了 `GROUPING` 函数。当在 `SELECT`、`HAVING`、`ORDER BY` 子句中使用 `GROUPING` 函数时，TiDB 会在逻辑优化阶段对其进行改写，将 `GROUPING` 函数与分组项（`GROUP BY` items）之间的关系，转化为与维度分组计算逻辑有关的 `GID`，并将此 `GID` 以 metadata 形式填充到新的 `GROUPING` 函数当中。
```

例如：

```sql
SELECT count(1) FROM t GROUP BY a,b,c WITH ROLLUP;
```

在此示例中，count(1) 的计算结果将分别在 {a,b,c}、{a,b}、{a}、{} 一共 4 个分组上进行聚合，然后输出各分组的汇总数据。

> **注意：**
>
> TiDB 暂不支持 Cube 语法。

## 使用场景

多列数据的聚合汇总输出一般常用于 OLAP（Online Analytical Processing）场景。通过使用 `WITH ROLLUP` 修饰符，你可以在聚合结果中得到额外的行，以展示不同维度的汇总信息，从而实现高级的数据分析和报表生成。

## 准备条件

目前，TiDB 仅在 TiFlash MPP 模式下支持为 `WITH ROLLUP` 语法生成有效的执行计划，因此你的 TiDB 集群需要包含 TiFlash 节点，并且对目标分析表进行了正确的 TiFlash 副本的配置。

更多信息，请参考[扩容 TiFlash 节点](/scale-tidb-using-tiup.md#扩容-tiflash-节点)。

## 使用示例

假如有一张名为 `bank` 的银行利润表，包含年（`year`）、月（`month`）、日（`day`）和利润（`profit`）列。

```sql
CREATE TABLE bank
(
    year    INT,
    month   VARCHAR(32),
    day     INT,
    profit  DECIMAL(13, 7)
);

ALTER TABLE bank SET TIFLASH REPLICA 1; -- 为该表添加一个 TiFlash 副本

INSERT INTO bank VALUES(2000, "Jan", 1, 10.3),(2001, "Feb", 2, 22.4),(2000,"Mar", 3, 31.6)
```

如需查看银行每年的利润，可以用一个简单的 `GROUP BY` 的子句来实现：

```sql
SELECT year, SUM(profit) AS profit FROM bank GROUP BY year;
+------+--------------------+
| year | profit             |
+------+--------------------+
| 2001 | 22.399999618530273 |
| 2000 |  41.90000057220459 |
+------+--------------------+
2 rows in set (0.15 sec)
```

> 对于银行报表来说，除了每年的利润之外，通常还需要计算所有年份的总利润和每个月的总利润，以进行更高层次或更详细的利润分析。在 v7.4.0 之前的版本中，你需要在多个查询中使用不同的 `GROUP BY` 子句，并将结果使用 UNION 连接，才能得到聚合汇总的结果。从 v7.4.0 起，你可以直接在单个查询的 `GROUP BY` 子句中添加 `WITH ROLLUP` 修饰符，即可得到所需的结果：

Reviewer 2: ```suggestion
对于银行报表来说，除了每年的利润之外，通常还需要计算所有年份的总利润或每个月的总利润，以进行更高层次或更详细的利润分析。在 v7.4.0 之前的版本中，你需要在多个查询中使用不同的 `GROUP BY` 子句，并将结果使用 UNION 连接，才能得到聚合汇总的结果。从 v7.4.0 起，你可以直接在单个查询的 `GROUP BY` 子句中添加 `WITH ROLLUP` 修饰符，即可得到所需的结果：
```

```sql
SELECT year, month, SUM(profit) AS profit from bank GROUP BY year, month WITH ROLLUP ORDER BY year desc, month desc;
+------+-------+--------------------+
| year | month | profit             |
+------+-------+--------------------+
| 2001 | Feb   | 22.399999618530273 |
| 2001 | NULL  | 22.399999618530273 |
| 2000 | Mar   | 31.600000381469727 |
| 2000 | Jan   | 10.300000190734863 |
| 2000 | NULL  |  41.90000057220459 |
| NULL | NULL  |  64.30000019073486 |
+------+-------+--------------------+
6 rows in set (0.025 sec)
```

> 以上结果包含了按照年份、月份、以及整体所有维度的聚合数据。其中，未出现 `NULL` 值的行表示该行 `profit` 是按照年份和月份分组计算的结果，`month` 列的 `NULL` 值表示该行 `profit` 是按照月份聚合计算的结果，`year` 列的 `NULL` 值表示该行 `profit` 是按照年份聚合计算的结果。

Reviewer 2: ```suggestion
以上结果包含了按照年份和月份、按照年份、以及整体所有维度的聚合数据。其中，未出现 `NULL` 值的行表示该行 `profit` 是按照年份和月份分组计算的结果，`month` 列的 `NULL` 值表示该行 `profit` 是按照月份聚合计算的结果，`year` 列的 `NULL` 值表示该行 `profit` 是按照年份聚合计算的结果。
```
